Overhaul treatment of nonbonded exceptions

andrrizzi opened the pull request:
Continues #39 .

jchodera committed:
Update binstar badge to anaconda.org

jchodera committed:
Bump version to 1.3.0

jchodera committed:
Rework (and hopefully fix) decoupling implementation; add tests for annihilated energy terms.

jchodera committed:
Lots of bugfixes to AbsoluteAlchemicalFactory and tests, though tests don't pass yet.

jchodera committed:
More debugging

jchodera committed:
More debugging.

jchodera committed:
Start building testing framework to compare energy component-by-component with alchemically-modified system

jchodera committed:
Merge remote-tracking branch 'upstream/master' into fix-exceptions

andrrizzi committed:
Fix typo

andrrizzi committed:
Fix test typo

andrrizzi committed:
Generalized exception treatment to handle alchemical-nonalchemical decoupled exceptions

andrrizzi committed:
Fix exclusions in CustomNonbondedForces

andrrizzi commented:
Ok, I think I've found a way to handle correctly also exceptions between alchemical atoms and non-alchemical ones for decoupling.

In order to do this I had to treat the decoupled intra-alchemical interactions [as exceptions in the unmodified nonbonded force](https://github.com/choderalab/alchemy/blob/fix-exceptions/alchemy/alchemy.py#L1086-L1115).

This is a summary table showing how contributions are now distributed across forces

```
# ------------------------------------------------------------------------------------------------------------------------------------------------
# FORCE                                 | ANNIHILATED                                     | DECOUPLED                                            |
# ------------------------------------------------------------------------------------------------------------------------------------------------
# nonbonded_force (unmodified)          | all interactions nonalchemical-nonalchemical    | all interactions nonalchemical-nonalchemical         |
#                                       | all exceptions nonalchemical-nonalchemical      | all exceptions nonalchemical-nonalchemical           |
#                                       |                                                 | all interactions alchemical-alchemical               |
#                                       |                                                 | all exceptions alchemical-alchemical                 |
# ------------------------------------------------------------------------------------------------------------------------------------------------
# sterics_custom_nonbonded_force        | sterics interactions alchemical-allatoms        | sterics interactions alchemical-nonalchemical        |
# ------------------------------------------------------------------------------------------------------------------------------------------------
# electrostatics_custom_nonbonded_force | electrostatics interactions alchemical-allatoms | electrostatics interactions alchemical-nonalchemical |
# ------------------------------------------------------------------------------------------------------------------------------------------------
# sterics_custom_bond_force             | sterics exceptions alchemical-allatoms          | sterics exceptions alchemical-nonalchemical          |
# ------------------------------------------------------------------------------------------------------------------------------------------------
# electrostatics_custom_bond_force      | electrostatics exceptions alchemical-allatoms   | electrostatics exceptions alchemical-nonalchemical   |
# ------------------------------------------------------------------------------------------------------------------------------------------------
```

Now I reactivated the test `compareEnergies` which is failing. I'm now augumenting `check_interacting_energy_components()` test to dissect the contribution and understand where we are failing.

For some reason when I run this tests on my local laptop and on the cluster I get (sometimes) `Abort trap: 6`. I'll try to do reinstall conda from scratch and see if I can get rid of it since Travis doesn't complain.

jchodera commented:
> In order to do this I had to treat the decoupled intra-alchemical interactions as exceptions in the unmodified nonbonded force.

This sounds correct to me.

> This is a summary table showing how contributions are now distributed across forces

Great table!

> Now I reactivated the test compareEnergies which is failing. I'm now augumenting check_interacting_energy_components() test to dissect the contribution and understand where we are failing.

Fantastic. I think this is exactly what we need to be doing.

> For some reason when I run this tests on my local laptop and on the cluster I get (sometimes) Abort trap: 6. I'll try to do reinstall conda from scratch and see if I can get rid of it since Travis doesn't complain.

Are these energy comparison tests or actual dynamics simulations? This is likely the CPU platform running into a segfault.

andrrizzi commented:
> This is likely the CPU platform running into a segfault.

After reinstalling miniconda from scratch, I isolated the code giving me `Abort trap: 6` to [the line where we obtain the state](https://github.com/choderalab/alchemy/blob/fix-exceptions/alchemy/tests/test_alchemy.py#L465). But is seems to be quite random, it doesn't happen all the time (which luckly means I can keep debugging alchemy locally).

jchodera commented:
Does it abort if you just get the positions? Or is it the energy calculation?

If the positions are OK, you can serialize the system and state out to XML and then submit that as a bug report to openmm.

andrrizzi commented:
> If the positions are OK, you can serialize the system and state out to XML and then submit that as a bug.

Right! I didn't think about that, thanks. I think it's the energy. I'll open an issue on openmm soon.

andrrizzi committed:
Dissect components of nonbonded force to test correct distribution

andrrizzi committed:
Implemented consistent_exceptions option and fixed 1,4 electrostatic exceptions

andrrizzi committed:
Expand tests to dissect nonbonded force contribution also by atom groups

andrrizzi committed:
Rework NonbondedForce scheme to model decoupled electrostatics with correct functional form

andrrizzi committed:
Adapt tests to new scheme for nonbonded forces

andrrizzi commented:
I've reworked the implementation of nonbonded force with the scheme suggested by @Lnaden, and rewrote the test `check_interacting_energy_components()` to check all different contributions by atom group (non-alchemical, alchemical, interface) and type (sterics, electrostatics). It should be pretty comprehensive.

I tried to implement this in the clearest way possible and added a lot of comments, but let me know if you think there's something that can be improved/changed.

I'll now extend `check_interacting_energy_components()` to test also for other forces other than `NonbondedForce`.

andrrizzi committed:
Fix tests precision issue

andrrizzi committed:
Fix test platform argument

andrrizzi committed:
Adapt check_noninteracting_energy_components test to new Nonbonded scheme

andrrizzi committed:
Test also fully interacting forces other than Nonbonded

Lnaden commented on alchemy/alchemy.py:
```diff
@@ -884,24 +947,32 @@ def _alchemicallyModifyNonbondedForce(self, system, reference_force):
             if alpha_ewald == 0.0:
                 # If alpha is 0.0, alpha_ewald is computed by OpenMM from from the error tolerance.
                 [alpha_ewald, nx, ny, nz] = reference_force.getPMEParameters()
+                assert (alpha_ewald != 0.0) # we can't have alpha be zero
```
Isn't this check redundant?

andrrizzi replied:
Not sure why that assert is there (@jchodera?), but it looks like we can move the `assert` outside the if-clause?

Lnaden commented on alchemy/alchemy.py:
```diff
@@ -884,24 +947,32 @@ def _alchemicallyModifyNonbondedForce(self, system, reference_force):
+            raise Exception("Nonbonded method %s not supported yet." % str(nonbonded_method))
 
         # Add additional definitions common to all methods.
         sterics_energy_expression += "reff_sterics = sigma*((softcore_alpha*(1.-lambda_sterics)^softcore_b + (r/sigma)^softcore_c))^(1/softcore_c);" # effective softcore distance for sterics
```
This is really just a formatting issue, but can we declare floats as `1.0` instead of just `1.`? It can get a bit confusing from a human readable perspective.

andrrizzi replied:
Sure! I'll fix it with next commit.

Lnaden commented on alchemy/alchemy.py:
```diff
@@ -971,40 +1168,82 @@ def _alchemicallyModifyNonbondedForce(self, system, reference_force):
+
+        # Turn off all exception contributions from alchemical atoms in the NonbondedForce
+        # modelling non-alchemical atoms only
+        for exception_index in range(nonbonded_force.getNumExceptions()):
```
Could this loop be rolled in with the last one (minor optimization, not really important)?

andrrizzi replied:
I purposely left it on a second loop because I thought it would make it more readable, but I can include it in the previous for loop if you think it's better.

Lnaden reviewed (approved):
I have just a few small changes I would make but otherwise this looks great!

andrrizzi commented:
I'll now turn off the `PME` test cases fed to `check_interacting_energy_components()` since we don't expect them to work yet. I'll then clean up tests to be sure we have not left out anything while making temporary changes for debugging, and then I think this is ready to merge if all tests will pass.

jchodera commented:
> I'll now turn off the PME test cases fed to check_interacting_energy_components() since we don't expect them to work yet. I'll then clean up tests to be sure we have not left out anything while making temporary changes for debugging, and then I think this is ready to merge if all tests will pass.

Can we just disable the electrostatics part of the test? We should leave the test that checks alchemical overlap for PME.

andrrizzi commented:
> Can we just disable the electrostatics part of the test?

Yes! Even better, thanks.

andrrizzi committed:
Deactivate check particle electrostatics interactions for PME

andrrizzi committed:
Fixed precision in compute_energy test utility function

andrrizzi committed:
Removed unused variables and imports

andrrizzi committed:
Reactivated test cases turned off for debugging

andrrizzi committed:
Changed 1. to 1.0 in energy expression

andrrizzi committed:
Make Travis not run tests marked as slow

andrrizzi commented on alchemy/alchemy.py:
```diff
@@ -1262,6 +1541,51 @@ def perturbSystem(cls, system, alchemical_state):
+            import copy
+            system = copy.deepcopy(self.alchemically_modified_system)
+            # Separate all forces into separate force groups.
+            assert (system.getNumForces() < 16), "self.alchemically_modified_system has more than 16 force groups; can't compute individual force component energies."
```
@jchodera is there a particular reason why you set this to 16? One of the test cases have exactly 16 forces and the openmm docs says that legal values for `.getState()` go from 0 to 31.

Lnaden replied:
Taking a guess: 
16 forces, possibly all alchemically split between electrostatics and sterics, 32 possible forces (max number of `ForceGroup`s. However, we always need at least 1 group to drop all the non-alchemical forces, so we can only ever have 15 forces max.

andrrizzi commented:
One of the tests (Alanine dipeptide in vacuum) is failing because of the OpenMM bug that assign complete interaction when one of the two sets given to `addInteractionGroup(set1, set2)` is empty (pandegroup/openmm#1554).

Is it ok if I deactivate the test until the new OpenMM gets released?

andrrizzi committed:
Fix PME test and removed redundant compareSystemEnergy after checking interactions

jchodera commented:
> One of the tests (Alanine dipeptide in vacuum) is failing because of the OpenMM bug that assign complete interaction when one of the two sets given to addInteractionGroup(set1, set2) is empty (pandegroup/openmm#1554).
> Is it ok if I deactivate the test until the new OpenMM gets released?

Yes!

jchodera commented:
The 16 force limit was probably just my misunderstanding of the OpenMM limitation!

andrrizzi commented:
Great! I'll fix the last tests today and merge it if there are no objections.

Lnaden commented:
none from me

andrrizzi committed:
Fix limit force groups to 32

andrrizzi committed:
Temporarily disabled alanine dipeptide test case until OpenMM bugfix is released

andrrizzi committed:
Add log output to avoid travis timeout

andrrizzi committed:
Try again avoiding travis timeout

andrrizzi commented:
Hmm, [this Travis error](https://travis-ci.org/choderalab/alchemy/jobs/162456989#L862) doesn't look like a timeout and happens at random lines inside `check_interacting_energy_components` (judging from the `print` statements).

According to Travis docs, it is probably [exhausting the memory](https://docs.travis-ci.com/user/common-build-problems/#My-build-script-is-killed-without-any-error) which is consistent with the problem I've been having when contexts get continuously created and deleted.

Those tests pass locally on my machine so, if we are in a rush to merge this PR, I can temporarily deactivate them, see if Travis pass all other tests, and then reactivate them and merge. I'll try to debug the OpenMM problem over the weekend.

andrrizzi commented:
Just to give references to the issues I'm referring to: pandegroup/openmm#1588, and probably also choderalab/yank#397 is related to this.

jchodera commented on alchemy/tests/test_alchemy.py:
```diff
@@ -146,24 +146,51 @@ def write_file(filename, contents):
 
+
+def compute_energy_force(system, positions, force_name, platform=None, precision=None):
+    system = copy.deepcopy(system)  # Copy to avoid modifications
```
Maybe we can `del system` at the end of this function?

andrrizzi replied:
Good idea. Let me try.

jchodera commented:
Yeah, I'm not sure what to make of this error. I think we might need to wait for the memory leak in https://github.com/pandegroup/openmm/issues/1588 to be tracked down.

andrrizzi committed:
Delete copied system after computing energy

jchodera commented:
Maybe we should just mark the test as `slow` for now and be sure to test locally until the `openmm` memory leak is found.

andrrizzi commented:
Sounds good. Local tests pass on my machine, so if the other tests are successful on travis I'll merge this.

andrrizzi committed:
Set test_alchemical_accuracty as slow test

jchodera commented:
Woohoo! Tests pass!

andrrizzi commented:
Great! Merging.
